genesyscloud/outbound_ruleset/resource_genesyscloud_outbound_ruleset.go

```diff
@@ -131,3 +144,59 @@ func deleteOutboundRuleset(ctx context.Context, d *schema.ResourceData, meta int
 		return retry.RetryableError(fmt.Errorf("Outbound Ruleset %s still exists", d.Id()))
 	})
 }
+
+// filterOutboundRuleset filters rulesets by removing the ones that are referencing skills that no longer exist in GC
+func filterOutboundRulesets(ruleSets []platformclientv2.Ruleset, skillMap resourceExporter.ResourceIDMetaMap) ([]platformclientv2.Ruleset, diag.Diagnostics) {
+	var filteredRuleSets []platformclientv2.Ruleset
+	log.Printf("Filtering outbound rule sets")
+
+RuleSetLoop:
+	for _, ruleSet := range ruleSets {
+		for _, rule := range *ruleSet.Rules {
+			// look through rule actions to check if the referenced skills exist in our skill map or not
+			for _, action := range *rule.Actions {
+				if action.ActionTypeName != nil && strings.ToLower(*action.ActionTypeName) == "set_skills" && action.Properties != nil {
+					for id, value := range *action.Properties {
+						if strings.ToLower(id) == "skills" {
+							// the property value is a json string wrapping an array of skill ids, need to convert it back to a slice to check if each skill exists
+							var skillIds []string
+							err := json.Unmarshal([]byte(value), &skillIds)
+							if err != nil {
+								log.Printf("Error decoding JSON: %s", err)
+								return nil, diag.Errorf("Failed to filter ruleset: %s", err)
+							}
+							for _, skillId := range skillIds {
+								_, found := skillMap[skillId]
+								if !found { // skill id referenced by the rule action is not found in the skill map, we filter the ruleset out and evaluate the next one.
+									log.Printf("Removing ruleset %s, the skill %s used in action does not exist in GC anymore", *ruleSet.Id, skillId)
+									continue RuleSetLoop
+								}
+							}
+						}
+					}
+				}
+			}
+			// look through rule conditions to check if the referenced skills exist in our skill map or not
+			for _, condition := range *rule.Conditions {
+				if condition.AttributeName != nil && strings.ToLower(*condition.AttributeName) == "skill" {
+					if condition.Value != nil {
+						var found bool
+						for _, value := range skillMap {
+							if value.Name == *condition.Value {
+								found = true
+								break
+							}
+						}
+						if !found { // skill name referenced by rule condition is not found in the skill map, we filter the rulset out and evaluate the next one
+							log.Printf("Removing ruleset %s, the skill %s used in condition does not exist in GC anymore", *ruleSet.Id, *condition.Value)
+							continue RuleSetLoop
```

Comment:
Found this.  I don't know.  This feels a little dirty.  Its not a goto, but...... :)

Reply:
Agreed, got overly excited about labeled loops. Removed